Panic runtime error on form refresh after appending item

### Describe the bug:
after appending to form.items, form.refresh creates a runtime error
```
panic: runtime error: index out of range [2] with length 2

goroutine 42 [running]:
fyne.io/fyne/v2/widget.(*Form).updateLabels(0xc000156420)
        /home/benoit/go/pkg/mod/fyne.io/fyne/v2@v2.0.4/widget/form.go:210 +0x20e
```


### To Reproduce:
the following code reproduces the error:
```
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/data/validation"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	w := a.NewWindow("testing")

	typeSelect := widget.NewSelect([]string{"Source", "Processor", "Sink"},
		func(s string) {})

	nbInputsEntry := widget.NewEntry()
	nbInputsEntry.Validator = validation.NewRegexp(`\d`, "Must contain a number")

	form := widget.NewForm(widget.NewFormItem("node type", typeSelect))

	typeSelect.OnChanged = func(s string) {
		switch s {
		case "Source":

		case "Processor":
			form.Items = append(form.Items, widget.NewFormItem("number of inputs", nbInputsEntry))
			form.Refresh()
		case "Sink":
		}
	}

	w.SetContent(form)
	w.Resize(fyne.NewSize(300, 100))
	w.ShowAndRun()
}
```

### Device (please complete the following information):
 - **OS:**  Linux, Manjaro KDE
 - **Version:** 
 - **Go version:** 1.16.7
 - **Fyne version:** 2.0.4

--- a/widget/form.go
+++ b/widget/form.go
@@ -85,6 +85,7 @@ func (f *Form) MinSize() fyne.Size {
 // Refresh updates the widget state when requested.
 func (f *Form) Refresh() {
 	cache.Renderer(f.super()) // we are about to make changes to renderer created content... not great!
+	f.ensureRenderItems()
 	f.updateButtons()
 	f.updateLabels()
 	f.BaseWidget.Refresh()
@@ -193,6 +194,20 @@ func (f *Form) checkValidation(err error) {
 	}
 }
 
+func (f *Form) ensureRenderItems() {
+	objects := f.itemGrid.Objects
+	for i, item := range f.Items {
+		if i < len(objects)/2 {
+			continue
+		}
+
+		objects = append(objects, f.createLabel(item.Text))
+		f.setUpValidation(item.Widget, i)
+		objects = append(objects, f.createInput(item))
+	}
+	f.itemGrid.Objects = objects
+}
+
 func (f *Form) setUpValidation(widget fyne.CanvasObject, i int) {
 	updateValidation := func(err error) {
 		if err == errFormItemInitialState {
@@ -263,16 +278,9 @@ func (f *Form) CreateRenderer() fyne.WidgetRenderer {
 	buttons := &fyne.Container{Layout: layout.NewGridLayoutWithRows(1), Objects: []fyne.CanvasObject{f.cancelButton, f.submitButton}}
 	f.buttonBox = &fyne.Container{Layout: layout.NewBorderLayout(nil, nil, nil, buttons), Objects: []fyne.CanvasObject{buttons}}
 
-	objects := make([]fyne.CanvasObject, len(f.Items)*2)
-	for i, item := range f.Items {
-		objects[i*2] = f.createLabel(item.Text)
-
-		f.setUpValidation(item.Widget, i)
-		objects[i*2+1] = f.createInput(item)
-	}
-	f.itemGrid = fyne.NewContainerWithLayout(layout.NewFormLayout(), objects...)
-
+	f.itemGrid = &fyne.Container{Layout: layout.NewFormLayout()}
 	renderer := NewSimpleRenderer(fyne.NewContainerWithLayout(layout.NewVBoxLayout(), f.itemGrid, f.buttonBox))
+	f.ensureRenderItems()
 	f.updateButtons()
 	f.updateLabels()
 	f.checkValidation(nil) // will trigger a validation check for correct intial validation status

--- a/widget/form.go
+++ b/widget/form.go
@@ -195,17 +195,27 @@ func (f *Form) checkValidation(err error) {
 }
 
 func (f *Form) ensureRenderItems() {
-	objects := f.itemGrid.Objects
+	done := len(f.itemGrid.Objects) / 2
+	if done >= len(f.Items) {
+		f.itemGrid.Objects = f.itemGrid.Objects[0 : len(f.Items)*2]
+		return
+	}
+
+	adding := len(f.Items) - done
+	objects := make([]fyne.CanvasObject, adding*2)
+	off := 0
 	for i, item := range f.Items {
-		if i < len(objects)/2 {
+		if i < done {
 			continue
 		}
 
-		objects = append(objects, f.createLabel(item.Text))
+		objects[off] = f.createLabel(item.Text)
+		off++
 		f.setUpValidation(item.Widget, i)
-		objects = append(objects, f.createInput(item))
+		objects[off] = f.createInput(item)
+		off++
 	}
-	f.itemGrid.Objects = objects
+	f.itemGrid.Objects = append(f.itemGrid.Objects, objects...)
 }
 
 func (f *Form) setUpValidation(widget fyne.CanvasObject, i int) {

--- a/widget/form_test.go
+++ b/widget/form_test.go
@@ -44,6 +44,19 @@ func TestForm_Append(t *testing.T) {
 	assert.Equal(t, item, form.Items[2])
 }
 
+func TestForm_Append_Items(t *testing.T) {
+	form := &Form{Items: []*FormItem{{Text: "test1", Widget: NewEntry()}}}
+	assert.Equal(t, 1, len(form.Items))
+	renderer := test.WidgetRenderer(form)
+
+	form.Items = append(form.Items, NewFormItem("test2", NewEntry()))
+	assert.True(t, len(form.Items) == 2)
+
+	form.Refresh()
+	c := renderer.Objects()[0].(*fyne.Container).Objects[0].(*fyne.Container)
+	assert.Equal(t, "test2", c.Objects[2].(*canvas.Text).Text)
+}
+
 func TestForm_CustomButtonsText(t *testing.T) {
 	form := &Form{OnSubmit: func() {}, OnCancel: func() {}}
 	form.Append("test", NewEntry())

